pkg/engine/planner/physical/optimizer.go

```diff
@@ -0,0 +1,86 @@
+package physical
+
+import "github.com/grafana/loki/v3/pkg/engine/internal/types"
+
+type optimizer struct {
+	plan *Plan
+}
+
+var _ Visitor = (*optimizer)(nil)
+
+// VisitDataObjScan implements Visitor.
+func (o *optimizer) VisitDataObjScan(*DataObjScan) error {
+	return nil
+}
+
+// VisitFilter implements Visitor.
+func (o *optimizer) VisitFilter(node *Filter) error {
+	for i := 0; i < len(node.Predicates); i++ {
+		if ok := o.applyPredicatePushdown(node, node.Predicates[i]); ok {
+			// remove predicates that have been pushed down
+			node.Predicates = append(node.Predicates[:i], node.Predicates[i+1:]...)
+			i--
+		}
+	}
+	return nil
+}
+
+// VisitLimit implements Visitor.
+func (o *optimizer) VisitLimit(node *Limit) error {
+	o.applyLimitPushdown(node, node.Limit)
+	return nil
+}
+
+// VisitProjection implements Visitor.
+func (o *optimizer) VisitProjection(*Projection) error {
+	return nil
+}
+
+// VisitSortMerge implements Visitor.
+func (o *optimizer) VisitSortMerge(*SortMerge) error {
+	return nil
+}
+
+func (o *optimizer) applyLimitPushdown(node Node, limit uint32) bool {
```

Comment:
Technically this mixes visitor pattern with pattern matching. I'd use `accept` with a new visitor that returns a copy of the plan.

Reply:
Visitors are very clunky, when you only need to visit a single node. I made an implementation that uses visitors only https://github.com/grafana/loki/compare/chaudum/physical-plan-optimizer-visitor-pattern?expand=1#diff-2a71283093168d3d8e9260cb06603c22d45fe87ed20de3b0cb287a0e786a2410